File: hooks/combobox.ts
import * as combobox from "@zag-js/combobox";
import type { Collection } from "@zag-js/collection";
import { getAttributes, restoreAttributes, normalizeProps, renderPart, spreadProps, getBooleanOption, getOption } from "./util";
import { Component } from "./component";
import type { ViewHook } from "phoenix_live_view";
import type { Machine } from "@zag-js/core";

type Item = { value: string; label: string };
type InputBehavior = "autocomplete" | "autohighlight" | "none" | undefined;
type SelectionBehavior = "clear" | "replace" | "preserve" | undefined;

class Combobox extends Component<combobox.Context, combobox.Api> {
  initService(context: combobox.Context): Machine<any, any, any> {
    return combobox.machine(context);
  }

  initApi() {
    return combobox.connect(this.service.state, this.service.send, normalizeProps);
  }

  render() {
    const parts = ["root", "label", "control", "input", "trigger", "positioner", "content"];
    for (const part of parts) renderPart(this.el, part, this.api);
    this.renderItems();
  }

  renderItems() {
    for (const item of this.el.querySelectorAll<HTMLElement>("[data-part='item']")) {
      const value = item.dataset.value;
      const label = item.dataset.label;
      if (!value || !label) {
        console.error("Missing `data-value` or `data-label` attribute on item.");
        return;
      }

      spreadProps(item, this.api.getItemProps({ item: { value, label } }));
    }
  }
}

export interface ComboboxHook extends ViewHook {
  state: any;
  combobox: Combobox;
  attributeCache: any;
  items(): Item[];
  collection(): Collection<Item>;
  context(): combobox.Context;
}

export default {
  mounted() {
    this.combobox = new Combobox(this.el, this.context());
    this.combobox.init();
  },

  beforeUpdate() {
    const parts = ["root", "label", "control", "input", "trigger", "positioner", "content"];
    this.attributeCache = parts.map((part) => {
      return getAttributes(this.el, part);
    });
  },

  updated() {
    this.combobox.api.setCollection(this.collection());
    this.combobox.render();
    restoreAttributes(this.el, this.attributeCache);
  },

  beforeDestroy() {
    this.combobox.destroy();
  },

  items(): Item[] {
    return Array.from(this.el.querySelectorAll<HTMLElement>("[data-part='item']"))
      .map((item: HTMLElement) => {
        const value = item.dataset.value;
        const label = item.dataset.label;
        if (!value || !label) {
          console.error("Missing `data-value` or `data-label` attribute on item.");
          return;
        }

        return { value, label };
      })
      .filter((value) => value !== undefined) as Item[];
  },

  collection() {
    return combobox.collection({
      items: this.items(),
      itemToValue: (item: Item) => item.value,
      itemToString: (item: Item) => item.label,
    });
  },

  context(): combobox.Context {
    return {
      id: this.el.id,
      name: this.el.dataset.name,
      collection: this.collection(),
      inputBehavior: getOption(this.el, "inputBehavior", ["autocomplete", "autohighlight", "none"]) as InputBehavior,
      selectionBehavior: getOption(this.el, "selectionBehavior", ["clear", "replace", "preserve"]) as SelectionBehavior,
      multiple: getBooleanOption(this.el, "multiple"),
      disabled: getBooleanOption(this.el, "disabled"),
      readOnly: getBooleanOption(this.el, "readOnly"),
      loopFocus: getBooleanOption(this.el, "loopFocus"),
      allowCustomValue: getBooleanOption(this.el, "allowCustomValue"),
      onOpenChange: (details: combobox.OpenChangeDetails) => {
        if (this.el.dataset.onOpenChange) {
          this.pushEvent(this.el.dataset.onOpenChange, details);
        }
      },
      onInputValueChange: (details: combobox.InputValueChangeDetails) => {
        if (this.el.dataset.onInputValueChange) {
          this.pushEvent(this.el.dataset.onInputValueChange, details);
        }
      },
      onHighlightChange: (details: combobox.HighlightChangeDetails) => {
        if (this.el.dataset.onHighlightChange) {
          this.pushEvent(this.el.dataset.onHighlightChange, details);
        }
      },
      onValueChange: (details: combobox.ValueChangeDetails) => {
        if (this.el.dataset.onValueChange) {
          this.pushEvent(this.el.dataset.onValueChange, details);
        }
      },
    };
  },
} as ComboboxHook;

File: lib/turboprop/headless/combobox.ex
defmodule Turboprop.Headless.Combobox do
  @moduledoc false
  use Phoenix.Component

  import Turboprop.Headless

  attr :id, :string, default: nil
  attr :as, :any, default: "div"

  attr :name, :string, default: nil
  attr :input_behavior, :string, values: ["autohighlight", "autocomplete", "none"], default: "none"
  attr :selection_behavior, :string, values: ["clear", "replace", "preserve"], default: "clear"
  attr :disabled, :boolean, default: false
  attr :multiple, :boolean, default: false
  attr :loop_focus, :boolean, default: false
  attr :allow_custom_value, :boolean, default: false
  attr :read_only, :boolean, default: false

  attr :on_open_change, :string, default: nil
  attr :on_input_value_change, :string, default: nil
  attr :on_highlight_change, :string, default: nil
  attr :on_value_change, :string, default: nil

  attr :rest, :global
  slot :inner_block

  def combobox(assigns) do
    {name, assigns} = Map.pop(assigns, :name)
    {input_behavior, assigns} = Map.pop(assigns, :input_behavior)
    {selection_behavior, assigns} = Map.pop(assigns, :selection_behavior)
    {disabled, assigns} = Map.pop(assigns, :disabled)
    {multiple, assigns} = Map.pop(assigns, :multiple)
    {loop_focus, assigns} = Map.pop(assigns, :loop_focus)
    {allow_custom_value, assigns} = Map.pop(assigns, :allow_custom_value)
    {read_only, assigns} = Map.pop(assigns, :read_only)

    {on_open_change, assigns} = Map.pop(assigns, :on_open_change)
    {on_input_value_change, assigns} = Map.pop(assigns, :on_input_value_change)
    {on_highlight_change, assigns} = Map.pop(assigns, :on_highlight_change)
    {on_value_change, assigns} = Map.pop(assigns, :on_value_change)

    render_as_tag_or_component(assigns, %{
      "id" => assigns.id || id(),
      "phx-hook" => "Combobox",
      "data-name" => name,
      "data-input-behavior" => input_behavior,
      "data-selection-behavior" => selection_behavior,
      "data-disabled" => disabled,
      "data-multiple" => multiple,
      "data-loop-focus" => loop_focus,
      "data-allow-custom-value" => allow_custom_value,
      "data-read-only" => read_only,
      "data-on-open-change" => on_open_change,
      "data-on-input-value-change" => on_input_value_change,
      "data-on-highlight-change" => on_highlight_change,
      "data-on-value-change" => on_value_change
    })
  end

  attr :as, :any, default: "div"
  attr :rest, :global
  slot :inner_block

  def root(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "root"})
  end

  attr :as, :any, default: "div"
  attr :rest, :global
  slot :inner_block

  def control(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "control"})
  end

  attr :as, :any, default: "input"
  attr :rest, :global
  slot :inner_block

  def input(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "input"})
  end

  attr :as, :any, default: "button"
  attr :rest, :global
  slot :inner_block

  def trigger(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "trigger"})
  end

  attr :as, :any, default: "div"
  attr :rest, :global
  slot :inner_block

  def positioner(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "positioner"})
  end

  attr :as, :any, default: "ul"
  attr :rest, :global
  slot :inner_block

  def content(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "content"})
  end

  attr :id, :string, default: nil
  attr :as, :any, default: "li"
  attr :value, :string, required: true
  attr :label, :string, required: true
  attr :rest, :global
  slot :inner_block

  def item(assigns) do
    {value, assigns} = Map.pop(assigns, :value)
    {label, assigns} = Map.pop(assigns, :label)

    render_as_tag_or_component(assigns, %{
      "id" => assigns.id || id(),
      "data-part" => "item",
      "data-value" => value,
      "data-label" => label
    })
  end

  attr :as, :any, default: "label"
  attr :rest, :global
  slot :inner_block

  def label(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "label"})
  end
end
